lib/fd.c

```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: 2024, Skyler Ferrante <sjf5462@rit.edu>
+// SPDX-License-Identifier: BSD-3-Clause
+
+/**
+ * To protect against file descriptor omission attacks, we open the std file
+ * descriptors with /dev/null if they are not already open. Code is based on
+ * fix_fds from sudo.c.
+ */
+
+#include "prototypes.h"
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+void check_fds(void)
+{
+	int miss[3];
+
+	/**
+	 * Make sure stdin, stdout, stderr are open
+	 * If they are closed, set them to /dev/null
+	 */
+	miss[STDIN_FILENO] = fcntl(STDIN_FILENO, F_GETFL, 0) == -1;
+	miss[STDOUT_FILENO] = fcntl(STDOUT_FILENO, F_GETFL, 0) == -1;
+	miss[STDERR_FILENO] = fcntl(STDERR_FILENO, F_GETFL, 0) == -1;
+
+	if(miss[STDIN_FILENO] || miss[STDOUT_FILENO] || miss[STDERR_FILENO]) {
+		int devnull =
+			open ("/dev/null", O_RDWR, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
+
+		if (devnull == -1)
+			abort ();
+
+		if (miss[STDIN_FILENO] && dup2 (devnull, STDIN_FILENO) == -1)
+			abort ();
+
+		if (miss[STDOUT_FILENO] && dup2 (devnull, STDOUT_FILENO) == -1)
+			abort ();
+
+		if (miss[STDERR_FILENO] && dup2 (devnull, STDERR_FILENO) == -1)
+			abort ();
+
+		if (devnull > STDERR_FILENO)
+			close (devnull);
+	}
+}
```

Comment:
Using a helper function simplifies it a little bit.  How about this?

```c
static void check_fd(int fd);


void
check_fds(void)
{
	check_fd(STDIN_FILENO);
	check_fd(STDOUT_FILENO);
	check_fd(STDERR_FILENO);
}


static void
check_fd(int fd)
{
	int  devnull;

	if (fcntl(fd, F_GETFL, 0) != -1)
		return;

	devnull = open("/dev/null", O_RDWR, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
	if (devnull == -1)
		abort();

	if (dup2(devnull, fd) == -1)
		abort();

	if (devnull > STDERR_FILENO)
		close(devnull);
}
```

Reply:
I agree, this does look better. We can also remove the final if statement. The only downside is it it opens "/dev/null" a maximum of three times, while the original only opened it once (not that it really matters).

Reply:
While having fd leaks for /dev/null is probably not a problem, I'd prefer to avoid them.  Static analyzers may also start throwing warnings some day about it.

Mind that the if is there because a leak for fd 0 1 or 2 is not really a leak, since those fds we actually want them open.

Reply:
> (Alternatively, you might unconditionally close devnull.)

I think this is dangerous, since devnull is guaranteed to be equal to the fd we are checking. If stdin (0) is not open and we open(/dev/null) devnull should be equal to 0. The dup2 isn't actually necessary, its just a good idea to do, since a system might not hand out new fds starting at the lowest available.

> Mind that the if is there because a leak for fd 0 1 or 2 is not really a leak, since those fds we actually want them open.

It should only be a leak if the system doesn't give new fds starting at the lowest option available since we start at 0 and go to 3. But yeah, it's still probably a good idea to check.